pkg/server/controller.go

```diff
@@ -145,6 +282,264 @@ type indexPage struct {
 	BaseURL       string
 }
 
+func (ctrl *Controller) decodeGoogleCallbackResponse(resp *http.Response) (name string, err error) {
+	type callbackResponse struct {
+		ID            string
+		Email         string
+		VerifiedEmail bool
+		Picture       string
+	}
+
+	var userProfile callbackResponse
+	err = json.NewDecoder(resp.Body).Decode(&userProfile)
+	if err != nil {
+		return
+	}
+
+	name = userProfile.Email
+	return
+}
+
+func (ctrl *Controller) decodeGithubCallbackResponse(resp *http.Response) (name string, err error) {
+	type callbackResponse struct {
+		ID        int64
+		Email     string
+		Login     string
+		AvatarURL string
+	}
+
+	var userProfile callbackResponse
+	err = json.NewDecoder(resp.Body).Decode(&userProfile)
+	if err != nil {
+		return
+	}
+
+	name = userProfile.Login
+	return
+}
+
+func (ctrl *Controller) decodeGitLabCallbackResponse(resp *http.Response) (name string, err error) {
+	type callbackResponse struct {
+		ID        int64
+		Email     string
+		Username  string
+		AvatarURL string
+	}
+
+	var userProfile callbackResponse
+	err = json.NewDecoder(resp.Body).Decode(&userProfile)
+	if err != nil {
+		return
+	}
+
+	name = userProfile.Username
+	return
+}
+
+func (ctrl *Controller) callbacRedirectkHandler(getAccountInfoURL string, oauthConf *oauth2.Config, decodeResponse func(*http.Response) (string, error)) http.HandlerFunc {
+	return func(w http.ResponseWriter, r *http.Request) {
+		cookie, err := r.Cookie(stateCookieName)
+		if err != nil {
+			ctrl.httpServer.ErrorLog.Printf("There seems to be problem with state cookie: %v", err)
```

Comment:
The only returned error by `Cookie` is `ErrNoCookie`, this annotation may not be necessary.